phase2/src/phase2.rs

```
@@ -0,0 +1,100 @@
+use itertools::Itertools;
+
+use std::fs::File;
+use std::fs::OpenOptions;
+
+use parameters::*;
+use utils::Phase2Error;
+
+use crate::utils::map_io_phase2_error;
+
+/// Make a new contribution with entropy to the last params file (in_params_filename) 
+/// and writes to the new params file (out_params_filename).
+/// Returns the contribution hash if successful, otherwises returns the error message.
+pub fn contribute(
+    in_params_filename: &String,
+    out_params_filename: &String,
+    entropy: &String,
+    print_progress: bool,
+    progress_update_interval: u32,
+) -> Result<[u8; 64], Phase2Error> {
+    let disallow_points_at_infinity = false;
+
+    if print_progress {
+        println!("starting");
+    }
+    // Create an RNG based on a mixture of system randomness and user provided randomness
+    let mut rng = {
+        use byteorder::{ReadBytesExt, BigEndian};
+        use blake2::{Blake2b, Digest};
+        use rand::{SeedableRng, Rng, OsRng};
+        use rand::chacha::ChaChaRng;
+
+        let h = {
+            let mut system_rng = OsRng::new().unwrap();
+            let mut h = Blake2b::default();
+
+            // Gather 1024 bytes of entropy from the system
+            for _ in 0..1024 {
+                let r: u8 = system_rng.gen();
+                h.input(&[r]);
+            }
+
+            // Hash it all up to make a seed
+            h.input(&entropy.as_bytes());
+            h.result()
+        };
+
+        let mut digest = &h[..];
+
+        // Interpret the first 32 bytes of the digest as 8 32-bit words
+        let mut seed = [0u32; 8];
+        for i in 0..8 {
+            seed[i] = digest.read_u32::<BigEndian>().map_err(map_io_phase2_error)?;
+        }
+
+        ChaChaRng::from_seed(&seed)
+    };
+
+    let reader = OpenOptions::new()
+                            .read(true)
+                            .open(in_params_filename)
+                            .expect("unable to open.");
```

This will panic if the file does not exist. I think we'd rather return a meaningful error message instead.

---

Fixed in [new commit](https://github.com/MystenLabs/phase2-bn254/pull/1/commits/552cc6f9d54e4401585c386b0e1e5c3a9f9ee5c9#diff-5159906df3e74880b7c76cd21b4465665f1056e4c86b7bbc42cbeebb01ea7863R62)